astropy/coordinates/erfa_astrom.py

```diff
@@ -0,0 +1,260 @@
+    for each individual timestamp but interpolates linearly
+    between support points.
+
+    For the interpolation, float64 MJD values are used, so time precision
+    for the interpolation will be around a microsecond.
+
+    This can dramatically speed up coordinate transformations,
+    e.g. between CIRS and ICRS,
+    when obstime is an array of many values (factors of 10 to > 100 depending
+    on the selected resolution, number of points and the time range of the values).
+
+    The precision of the transformation will still be in the order of microseconds
+    for reasonable values of time_resolution, e.g. ``300 * u.s``.
+
+    Users should benchmark performance and accuracy with the default transformation
+    for their specific use case and then choose a suitable ``time_resolution``
+    from there.
+
+    This class is intended be used together with the ``erfa_astrom`` science state,
+    e.g. in a context manager like this
+
+    Example
+    -------
+    >>> from astropy.coordinates import SkyCoord, CIRS
+    >>> from astropy.coordinates.erfa_astrom import erfa_astrom, ErfaAstromInterpolator
+    >>> import astropy.units as u
+    >>> from astropy.time import Time
+    >>> import numpy as np
+
+    >>> obstime = Time.now() + np.linspace(0, 4, 1000) * u.hour
+    >>> with erfa_astrom.set(ErfaAstromInterpolator(300 * u.s)):
+    ...    cirs = SkyCoord.from_name('Crab').transform_to(CIRS(obstime=obstime))  # doctest: +REMOTE_DATA
+    '''
+
+    @u.quantity_input(time_resolution=u.day)
+    def __init__(self, time_resolution):
+        if time_resolution.to_value(u.ms) < 1:
+            warnings.warn(
+                f'Using {self.__class__.__name__} with `time_resolution`'
+                ' below 1ms might not improve performance',
+                AstropyWarning
+            )
+        if time_resolution.to_value(u.us) < 10:
+            warnings.warn(
+                f'Using {self.__class__.__name__} with `time_resolution`'
+                ' below 10 us is probably not practical since float64 for MJD '
+                ' is used for interpolation',
+                AstropyWarning
+            )
+        self.mjd_resolution = time_resolution.to_value(u.day)
+
+    def _get_support_points(self, obstime):
+        '''
+        Calculate support points for the interpolation.
+        This includes the minimum and maximum timestamp in obstime
+        to improve precision when the time resolution is wider than
+        the time range or only a single value is requested.
+        '''
+        mjd_scaled = np.ravel(obstime.mjd / self. mjd_resolution)
+
+        # ndmin=1 needed in case only a single coordinate is transformed
+        mjd_lower = np.array(np.floor(mjd_scaled), ndmin=1, copy=False)
+        mjd_upper = np.array(np.ceil(mjd_scaled), ndmin=1, copy=False)
+
+        # unique already does sorting
+        mjd_u = np.unique(np.concatenate([
+            [mjd_scaled.min(), mjd_scaled.max()],
```

Comment:
It's explained in the docstring. If the resolution is larger than the time range of the values, min and max provide better precision without harming performance.

Reply:
Personally, I'd rather remove the `[mjd_scaled.min(), mjd_scaled.max()]` as it will lead to surprising results in some cases. A minimal example:
```python
t1 = Time('2020-01-01T20:00:00') + [1, 2, 3, 4, 5] * u.day
t2 = Time('2020-01-01T20:00:00') + [2, 3, 4, 5, 6] * u.day
location = EarthLocation.of_site('Roque de los Muchachos')
crab = SkyCoord.from_name("Crab Nebula")

frame1 = AltAz(obstime=t1, location=location)
frame2 = AltAz(obstime=t2, location=location)

with erfa_astrom.set(ErfaAstromInterpolator(10 * u.day)):
    azel1 = crab.transform_to(frame1)
    azel2 = crab.transform_to(frame2)
    # printing array entries, which are present in both cases
    print(f'{azel1[1:].az}\n{azel1[1:].alt}')
    print(f'{azel2[:-1].az}\n{azel2[:-1].alt}')
# Would expect the same results for both, which is not the case:
# [82.93128862 83.326738   83.72438029 84.12447817] deg
# [36.87963601 37.73749477 38.59604886 39.45526105] deg
# [82.93128338 83.32673447 83.72437851 84.1244782 ] deg
# [36.87963684 37.73749523 38.59604895 39.45526079] deg
```
It is true that the min-max injection will lead to more precise results, but I would prefer to get the same results for same input times. At the moment, this depends on which other numbers are in the arrays.

Reply:
Does that mean you would also like to have the interpolation for single values? Right now, that uses the non-interpolating case.

Reply:
Good point. I'd say yes. If one doesn't want that behavior (interpolation even for single obstimes), one should not use the interpolated version anyways. But let's wait what others think about this.

Reply:
I'm a bit of two minds: ideally, the examples given should give the same answers, but it would also seem that if the code *knows* it is going to do more calculations for the interpolating case than for the regular case (as is certainly true for a single object), it might as well do the regular one. An in-between solution might be to always generate a complete MJD grid but discard items that for sure are never going to be used (i.e., that are not nearest right and left neighbours of requested MJD). That reasonably minimizes overhead to a factor 2 even in the worst case (single object or times separated by much more than the interpolation time), yet ensures the answers are always the same for a given interpolation time.

Reply:
Thanks for your opinion @mhvk. I think, what you suggest is already the current behavior, as the `ceil` and `floor` function would result in a "sparse" support if the input obstimes are not homogeneously distributed. Or am I misunderstanding something here? The only problem, which I see here is that the min and max injection "destroys" the regular spacing of the support grid for the elements that fall into the first and last grid cell.

Reply:
Ah, yes, I clearly had a different mental picture of the routine than was its reality! I think overall, I'd tend to agree to just remove the special-casing near the edge, and probably also the scalar case. Just keep the code as simple as possible.